### PR TITLE
[Compute] BREAKING CHANGE: Update NAT pool to NAT rule V2 for Standard LB SKU when creating VMSS

### PR DESCRIPTION
**Pull request by zhoxing-ms**
Feature request - Close: #22682

**Related command**

`az vmss create`

**Description**

**Testing Guide**

**History Notes**

[Compute] BREAKING CHANGE: `az vmss create`: Update NAT pool to NAT rule V2 for Standard LB SKU when creating VMSS
[Compute] `az vmss create`: Add new parameter `--nat-rule-name` to specify the name of NAT rule V2 when creating a new load balancer (NAT rule V2 is used to replace NAT pool)

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 1]
```diff
@@ -864,6 +864,36 @@ def build_load_balancer_resource(cmd, name, location, tags, backend_pool_name, n
...
+            "frontendPortRangeEnd": str(max(50119, 49999 + instance_count * (1 if disable_overprovision else 2))),
+            "protocol": "tcp",
+            "idleTimeoutInMinutes": 5
+        },
+        "dependsOn": [
+            "[concat('Microsoft.Network/loadBalancers/', '{}')]".format(lb_name)
+        ]
+    }
+
+    return nat_rule
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:894]: @fitzgeraldsteele @anavinahar Could you help confirm that the definition of NAT Rule V2 resource is in line with expectations?

**Comment by yonzhan** [comment]
Compute

**Comment by anavinahar** [comment]
@Mahip ***@***.***>, could you help look into this?


-a

________________________________

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 2]
```diff
@@ -867,6 +867,36 @@ def build_load_balancer_resource(cmd, name, location, tags, backend_pool_name, n
...
+        "properties": {
+            "frontendIPConfiguration": {
+                'id': "[concat({}, '/frontendIPConfigurations/', '{}')]".format(lb_id, frontend_ip_name)
+            },
+            "backendAddressPool": {
+                "id": "[concat({}, '/backendAddressPools/', '{}')]".format(lb_id, backend_pool_name)
+            },
+            "backendPort": backend_port,
+            "frontendPortRangeStart": "50000",
+            "frontendPortRangeEnd": str(max(50119, 49999 + instance_count * (1 if disable_overprovision else 2))),
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:888]: Would it be more intuitive to calculate this expression separately and use meaningful variable names?

Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:888]: Would it be more intuitive to calculate this expression separately and use meaningful variable names?

**Review by wangzelin007 — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py (hunk 2 above)
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:888]: This is a good suggestion. However, unfortunately, this code comes from legacy logic:
https://github.com/Azure/azure-cli/blob/ba4e12556c124cfbdd4d6f6e7409cd2a79f59f97/src/azure-cli/azure/cli/command_modules/vm/_template_builder.py#L827-L828
So I am not very clear about its original meaning. I may add some comments to explain it, what do you think of it?

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:888]:
```suggestion
            # This logic comes from the template of `inboundNatPools` to keep consistent with the frontend port of nat pool
            # keep 50119 as minimum for backward compat, and ensure over-provision is taken care of
            "frontendPortRangeEnd": str(max(50119, 49999 + instance_count * (1 if disable_overprovision else 2))),
```

Comment (yanzhudd) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:888]: Okay, looks great!

**Review by yanzhudd — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 3]
```diff
@@ -724,7 +724,8 @@ def load_arguments(self, _):
...
         c.argument('load_balancer_sku', resource_type=ResourceType.MGMT_NETWORK, min_api='2017-08-01', options_list=['--lb-sku'], arg_type=get_enum_type(LoadBalancerSkuName),
                    help="Sku of the Load Balancer to create. Default to 'Standard' when single placement group is turned off; otherwise, default to 'Basic'. The public IP is supported to be created on edge zone only when it is 'Standard'")
-        c.argument('nat_pool_name', help='Name to use for the NAT pool when creating a new load balancer.', options_list=['--lb-nat-pool-name', '--nat-pool-name'])
+        c.argument('nat_pool_name', help='Name to use for the NAT pool when creating a new load balancer.', options_list=['--lb-nat-pool-name', '--nat-pool-name'], deprecate_info=c.deprecate(target='--nat-pool-name', redirect='--nat-rule-name', hide=True))
+        c.argument('nat_rule_name', help='Name to use for the NAT rule v2 when creating a new load balancer. (it is used to replace nat pool)', options_list=['--lb-nat-rule-name', '--nat-rule-name'])
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py:728]:
```suggestion
        c.argument('nat_rule_name', help='Name to use for the NAT rule v2 when creating a new load balancer. (NAT rule V2 is used to replace NAT pool)', options_list=['--lb-nat-rule-name', '--nat-rule-name'])
```

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 4]
```diff
@@ -867,6 +867,38 @@ def build_load_balancer_resource(cmd, name, location, tags, backend_pool_name, n
...
+        "properties": {
+            "frontendIPConfiguration": {
+                'id': "[concat({}, '/frontendIPConfigurations/', '{}')]".format(lb_id, frontend_ip_name)
+            },
+            "backendAddressPool": {
+                "id": "[concat({}, '/backendAddressPools/', '{}')]".format(lb_id, backend_pool_name)
+            },
+            "backendPort": backend_port,
+            "frontendPortRangeStart": "50000",
+            # This logic comes from the template of `inboundNatPools` to keep consistent with the frontend port of nat pool
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:888]:
```suggestion
            # This logic comes from the template of `inboundNatPools` to keep consistent with the frontend port of NAT pool
```

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 5]
```diff
@@ -867,6 +867,38 @@ def build_load_balancer_resource(cmd, name, location, tags, backend_pool_name, n
...
+        "properties": {
+            "frontendIPConfiguration": {
+                'id': "[concat({}, '/frontendIPConfigurations/', '{}')]".format(lb_id, frontend_ip_name)
+            },
+            "backendAddressPool": {
+                "id": "[concat({}, '/backendAddressPools/', '{}')]".format(lb_id, backend_pool_name)
+            },
+            "backendPort": backend_port,
+            "frontendPortRangeStart": "50000",
+            # This logic comes from the template of `inboundNatPools` to keep consistent with the frontend port of NAT pool
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:888]:
```suggestion
            # This logic comes from the template of `inboundNatPools` to keep consistent with NAT pool
```

**Pull request merged**